retry/retry.go

```diff
@@ -30,61 +30,124 @@ type OnSuccessFunc func(elapsed time.Duration, attempt uint64, lastErr error)
+		fResultCh := make(chan error)
+
+		go func() {
+			fResultCh <- f(ctx)
+			close(fResultCh)
+		}()
+
+		select {
+		case err := <-fResultCh:
+			return err
+
+		case <-ctx.Done():
+			go func() {
+				// Drain channel and wait for it to be closed in the background.
+				for range fResultCh {
+				}
+			}()
+
+			return ctx.Err()
+		}
+	}
 }
 
-// WithBackoff retries the passed function if it fails and the error allows it to retry.
-// The specified backoff policy is used to determine how long to sleep between attempts.
+// WithBackoff retries the retryableFunc in case of failure.
+//
+// The passed ctx loosely restricts retries and prohibits retries if the context is done. It is also being passed to the
+// retryableFunc, which should honor the context as well. However, unless Settings.QuickContextExit is set, the
+// RetryableFunc can survive after the context has already finished.
+//
+// If retryableFunc has returned an error, retryable decides based on this error if another attempt should be made. This
+// package comes with the Retryable() function collecting common errors considered recoverable.
+//
+// For debouncing, a delay between a failed run and making another attempt can be defined by backoffFn.
+//
+// Additional configuration and tweaks can be set via settings, as further documented at the Settings type.
 func WithBackoff(
-	ctx context.Context, retryableFunc RetryableFunc, retryable IsRetryable, b backoff.Backoff, settings Settings,
+	ctx context.Context,
+	retryableFunc RetryableFunc,
+	retryable IsRetryable,
+	backoffFn backoff.Backoff,
+	settings Settings,
 ) (err error) {
 	// Channel for retry deadline, which is set to the channel of NewTimer() if a timeout is configured,
 	// otherwise nil, so that it blocks forever if there is no timeout.
 	var timeout <-chan time.Time
-
 	if settings.Timeout > 0 {
 		t := time.NewTimer(settings.Timeout)
 		defer t.Stop()
 		timeout = t.C
 	}
 
+	// funcWrapper is wrapped around potentially time-consuming blocks: the RetryableFunc and, if configured, the two
+	// callback functions. By default, funcWrapper is just a path-through identify function. However, with
+	// Settings.QuickContextExit set, it will be contextBoundFunc, directly bailing out when the context is done.
+	var funcWrapper = func(f func(context.Context) error) func(context.Context) error { return f }
+	if settings.QuickContextExit {
+		funcWrapper = contextBoundFunc
+	}
+
 	start := time.Now()
 	timedOut := false
-	for attempt := uint64(1); ; /* true */ attempt++ {
+	for attempt := uint64(1); ; attempt++ {
 		prevErr := err
 
-		if err = retryableFunc(ctx); err == nil {
-			if settings.OnSuccess != nil {
-				settings.OnSuccess(time.Since(start), attempt, prevErr)
+		err = funcWrapper(func(ctx context.Context) error {
+			err := retryableFunc(ctx)
+			if err == nil {
+				if settings.OnSuccess != nil {
+					settings.OnSuccess(time.Since(start), attempt, prevErr)
+				}
 			}
-
+			return err
+		})(ctx)
```

Comment:
Is the purpose of calling `OnSuccess` within the wrapped function so that this callback couldn't delay the return of this function either?

However, this seems to move calling the callbacks to another goroutine, which might be surprising and could introduce race conditions.

Reply:
Yes. Theoretically, both callback functions may block as well, while, unfortunately, they do not even receive a context to be bound to.

I am a bit unhappy with this architecture, but at least it does not break the API and needs a check for every `retry.WithBackoff` call. Another option would be to exclude the callbacks from the context-wrapper and stating this in the documentation. In most cases, the callbacks are only used for logging which should not block.